libs/agno/agno/compression/manager.py

```diff
@@ -1,195 +1,435 @@
                 ]
             )
-            return response.content
+
+            if not response.content:
+                return None
+
+            # Merge message IDs with previous compressed context
+            all_msg_ids = compressed_msg_ids
+            if compression_context and compression_context.message_ids:
+                all_msg_ids = all_msg_ids.union(compression_context.message_ids)
+
+            new_context = CompressedContext(
+                content=response.content,
+                message_ids=all_msg_ids,
+                updated_at=datetime.now(),
+            )
+
+            # Calculate and track token savings
+            original_tokens = count_text_tokens(conversation_text, self.model.id)
+            compressed_tokens = count_text_tokens(response.content, self.model.id)
+
+            self.stats["context_compressions"] = self.stats.get("context_compressions", 0) + 1
+            self.stats["messages_compressed"] = self.stats.get("messages_compressed", 0) + len(msgs_to_compress)
+            self.stats["original_context_tokens"] = self.stats.get("original_context_tokens", 0) + original_tokens
+            self.stats["compression_context_tokens"] = (
+                self.stats.get("compression_context_tokens", 0) + compressed_tokens
+            )
+
+            log_info(
+                f"Context compressed: {original_tokens} -> {compressed_tokens} tokens ({len(msgs_to_compress)} msgs)"
+            )
+
+            # 6. Rebuild message list: system + summary + current_user + last_tool_batch
+            new_messages: List[Message] = []
+
+            # Keep system messages
+            for msg in messages:
+                if msg.role == "system":
+                    new_messages.append(msg)
+                else:
+                    break
+
+            # Add summary of compressed messages as user message
+            summary_content = f"<previous_summary>\n{response.content}\n</previous_summary>"
+            summary_msg = Message(
+                role="user",
+                content=summary_content,
+                add_to_agent_memory=True,
+            )
+            new_messages.append(summary_msg)
+
+            # Add current user
+            new_messages.append(messages[current_user_idx])
+
+            # Add last tool batch (if exists)
+            if last_tool_idx is not None:
+                new_messages.extend(messages[last_tool_idx:])
+
+            # Replace messages in place
+            messages[:] = new_messages
+
+            return new_context
+
         except Exception as e:
-            log_error(f"Error compressing tool result: {e}")
-            return tool_content
+            log_error(f"Error compressing context: {e}")
 
-    def compress(self, messages: List[Message]) -> None:
-        """Compress uncompressed tool results"""
-        if not self.compress_tool_results:
-            return
+        return None
+
+    async def _acompress_context(
```

Comment:
Lets pull out the common parts into a helper. Only a small part of this is actually async.
Same for tool compression